server/mailet/mock-smtp-server/src/test/java/org/apache/james/mock/smtp/server/MockSMTPServerTest.java

```diff
@@ -184,6 +186,155 @@ void serverShouldReceiveMessageDataClient() throws Exception {
         }
     }
 
+    @Nested
+    class NumberOfAnswersTest {
+        private MockSMTPServer mockServer;
+        private FakeMail mail1;
+        private MimeMessage mimeMessage1;
+        private SMTPMessageSender smtpClient;
+        private SMTPBehaviorRepository behaviorRepository;
+
+        @BeforeEach
+        void setUp() throws Exception {
+            behaviorRepository = new SMTPBehaviorRepository();
+            mockServer = new MockSMTPServer(behaviorRepository);
+
+            mimeMessage1 = MimeMessageBuilder.mimeMessageBuilder()
+                .setSubject("test")
+                .setText("any text")
+                .build();
+            mail1 = FakeMail.builder()
+                .name("name")
+                .sender(BOB)
+                .recipients(ALICE, JACK)
+                .mimeMessage(mimeMessage1)
+                .build();
+
+            mockServer.start();
+            smtpClient = new SMTPMessageSender(DOMAIN)
+                .connect("localhost", mockServer.getPort());
+        }
+
+        @AfterEach
+        void tearDown() {
+            mockServer.stop();
+        }
+
+        @Test
+        void serverShouldKeepReceivingErrorResponseWhenAnytime() throws Exception {
+            behaviorRepository.setBehaviors(new MockSMTPBehavior(
+                MAIL_FROM,
+                Condition.MATCH_ALL,
+                Response.serverReject(SERVICE_NOT_AVAILABLE_421, "mock response"),
+                MockSMTPBehavior.NumberOfAnswersPolicy.anytime()));
+
+            try {
+                smtpClient.sendMessage(mail1);
+            } catch (SMTPConnectionClosedException e) {
+                // ignore first failure;
+            }
+
+            assertThatThrownBy(() -> smtpClient.sendMessage(mail1))
+                .isInstanceOf(SMTPConnectionClosedException.class)
+                .hasMessageContaining("421");
+        }
+
+        @Test
+        void serverShouldDecreaseNumberOfAnswerAfterMatched() throws Exception {
+            MockSMTPBehavior behavior = new MockSMTPBehavior(
+                MAIL_FROM,
+                Condition.MATCH_ALL,
+                Response.serverReject(SERVICE_NOT_AVAILABLE_421, "mock response"),
+                MockSMTPBehavior.NumberOfAnswersPolicy.times(5));
+
+            behaviorRepository.setBehaviors(behavior);
+
+            sendMessageIgnoreError();
+
+            assertThat(remainedAnswersOf(behavior))
+                .isEqualTo(4);
+        }
+
+        @Test
+        void serverShouldActLikeDefaultAfterGettingEnoughMatches() throws Exception {
+            MockSMTPBehavior behavior = new MockSMTPBehavior(
+                MAIL_FROM,
+                Condition.MATCH_ALL,
+                Response.serverReject(SERVICE_NOT_AVAILABLE_421, "mock response"),
+                MockSMTPBehavior.NumberOfAnswersPolicy.times(4));
+
+            behaviorRepository.setBehaviors(behavior);
+
+            sendMessageIgnoreError();
```

Comment:
I prefer a `for` and a variable

Reply:
I don't like `for`, not so many lines can save but harder code to read